### PR TITLE
Add declarationDir option

### LINKED ISSUE #6723: Add separate compiler option for generated declarations output directory.
I need to make an automated building of my typescript projects. Typescript compilation results (i.e .js and .d.ts files) need further separate processing. 

It would be nice if there is separate compilation option to specify output directory for generated declarations. Something like this:

```javascript
//tsconfig.json
{
  "CompilerOptions": {
    ...
      "declaration": true,
      "declarationDir": "target/typings"
    ...
  }
}```

### PR DESCRIPTION
**Pull request by masaeedu**
Fixes #6723.

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -2012,6 +2012,16 @@ namespace ts {
         return emitOutputFilePathWithoutExtension + extension;
     }
 
+    export function getDeclarationEmitOutputFilePath(sourceFile: SourceFile, host: EmitHost) {
+        const options = host.getCompilerOptions();
+        const outputDir = options.declarationDir || options.outDir; // Prefer declaration folder if specified
+        return options.declaration ? removeFileExtension(
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Turn this into an `if` and add a newline above.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -2012,6 +2012,16 @@ namespace ts {
         return emitOutputFilePathWithoutExtension + extension;
     }
 
+    export function getDeclarationEmitOutputFilePath(sourceFile: SourceFile, host: EmitHost) {
+        const options = host.getCompilerOptions();
+        const outputDir = options.declarationDir || options.outDir; // Prefer declaration folder if specified
+        return options.declaration ? removeFileExtension(
+            outputDir
+                ? getSourceFilePathInNewDir(sourceFile, host, outputDir)
+                : sourceFile.fileName
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Hoist this ternary into a local within the `if`

**Comment by DanielRosenwasser** [comment]
We have a function `verifyCompilerOptions` in `program.ts`. Can you add a line for `declarationDir` and `declaration` with `Diagnostics.Option_0_cannot_be_specified_without_specifying_option_1`?

**Comment by masaeedu** [comment]
@DanielRosenwasser Ok, done. I should mention that I wasn't sure what the expected behavior for declarationDir was when `outFile` or `out` are specified. Right now the behavior is to simply create a declaration file adjacent to the bundled JS file, regardless of whether the declarationDir option is specified. Possible alternative behaviors would be:

 - Produce an error when verifying options if declarationDir is specified alongside out or outFile
 - Produce the bundled declaration file in the path specified by declarationDir

Let me know if either of those are what's needed.

**Comment by DanielRosenwasser** [comment]
I think that if `--declarationDir` was specified separately from `--outFile` or `--out`, that was probably intentional and should be respected independently.

File: src/harness/harnessLanguageService.ts [hunk 3]
```diff
@@ -107,7 +107,7 @@ namespace Harness.LanguageService {
         }
     }
 
-    class DefaultHostCancellationToken implements ts.HostCancellationToken {
+    export class DefaultHostCancellationToken implements ts.HostCancellationToken {
```
Comment (mhegazy) [inline src/harness/harnessLanguageService.ts]: why do we need these exported?

File: src/compiler/utilities.ts [hunk 4]
```diff
@@ -2065,23 +2077,23 @@ namespace ts {
...
+                !isDeclarationFile(sourceFile)                                       // Not a declaration file
+                && (!isExternalModule(sourceFile) || !!getEmitModuleKind(options))); // and not a module, unless module emit enabled
+
             if (bundledSources.length) {
                 const jsFilePath = options.outFile || options.out;
                 const emitFileNames: EmitFileNames = {
                     jsFilePath,
                     sourceMapFilePath: getSourceMapFilePath(jsFilePath, options),
-                    declarationFilePath: getDeclarationEmitFilePath(jsFilePath, options)
+                    declarationFilePath: options.declaration ? removeFileExtension(jsFilePath) + ".d.ts" : undefined
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: since --declarationDir is ignored here, it should be an error to specify --out/--outFile + --declarationDir

Comment (masaeedu) [inline src/compiler/utilities.ts]: @DanielRosenwasser suggests that the `declarationDir` option be respected when specified alongside `out`/`outFile`. I think this means I should simply put the bundled declaration file in `declarationDir`.

File: src/harness/harnessLanguageService.ts (hunk 3 above)
Comment (masaeedu) [inline src/harness/harnessLanguageService.ts]: I can't reproduce this anymore, but I was getting compile errors concerning the visibility of these classes. I might have brainfarted and forgot to `jake local` before `jake tests`. I'll pull this commit out.

File: src/compiler/utilities.ts (hunk 4 above)
Comment (mhegazy) [inline src/compiler/utilities.ts]: i do not think this is correct. `declarationDir` should behave like `outDir`, we can add in the future a `declarationOutFile` to behave like `outFile`

**Comment by masaeedu** [comment]
Okay, I've pulled out the test harness changes and made specifying `declarationDir` alongside `out` or `outFile` invalid per @mhegazy's instructions

**Comment by masaeedu** [comment]
Don't really get where those linter errors are coming from. I haven't touched that file, and running the tests locally seems to pass. I think the offending whitespace was introduced with https://github.com/Microsoft/TypeScript/commit/54ae2709d342011af7d32036071a3f5365463f76 @mhegazy

**Comment by RyanCavanaugh** [comment]
Sending a PR to fix lint

File: src/compiler/program.ts [hunk 5]
```diff
@@ -1670,6 +1670,15 @@ namespace ts {
                 }
             }
 
+            if (options.declarationDir) {
+                if (!options.declaration) {
+                    programDiagnostics.add(createCompilerDiagnostic(Diagnostics.Option_0_cannot_be_specified_without_specifying_option_1, "declarationDir", "declaration"));
+                }
+                if (options.out || options.outFile) {
+                    programDiagnostics.add(createCompilerDiagnostic(Diagnostics.Option_0_cannot_be_specified_with_option_1, "declarationDir", options.out ? "out" : options.outFile));
```
Comment (mhegazy) [inline src/compiler/program.ts]: `"outFile"` instead of `options.outFile`

**Comment by mhegazy** [comment]
one small typo. other than that LGTM

**Comment by masaeedu** [comment]
@mhegazy Whoops. Fixed

**Comment by mhegazy** [comment]
Thanks!

**Pull request merged**